CactbotOverlay/CactbotEventSource.cs

```diff
@@ -256,146 +271,29 @@ public CactbotOverlay(CactbotOverlayConfig config)
       log_lines_semaphore_.Release();
     }
 
-    // This is called by the OverlayPlugin every 1s which is not often enough for us, so we
-    // do our own update mechanism as well.
-    protected override void Update() {
-      if (ffxiv_ == null)
-        Initialize();
-      SendSlowRateEvents();
-    }
-
     // Sends an event called |event_name| to javascript, with an event.detail that contains
     // the fields and values of the |detail| structure.
     public void DispatchToJS(JSEvent e) {
-      // Make sure we're not in the middle of shutdown.
-      if (this.Overlay.Renderer == null)
-        return;
-
-      // DispatchToJS can be called from multiple threads (both fast and main).
-      // OverlayMessage also calls this, which can be on other threads as well.
-      // Could consider adding per-thread builders in TLS to avoid a lock, but maybe overkill.
-      lock (this.dispatch_lock_) {
-        dispatch_string_builder_.Append("document.dispatchEvent(new CustomEvent('");
-        dispatch_string_builder_.Append(e.EventName());
-        dispatch_string_builder_.Append("', { detail: ");
-        dispatch_serializer_.Serialize(dispatch_json_writer_, e);
-        dispatch_string_builder_.Append(" }));");
-        this.Overlay.Renderer.ExecuteScript(dispatch_string_builder_.ToString());
-        dispatch_string_builder_.Clear();
-      }
-    }
-
-    // Events that we want to update less often because they aren't are critical.
-    private void SendSlowRateEvents() {
-      // NOTE: This function runs on a different thread that SendFastRateEvents(), so anything it calls needs to be thread-safe!
+      JObject ev = new JObject();
+      ev["type"] = e.EventName();
+      ev["detail"] = JObject.FromObject(e);
+      DispatchEvent(ev);
     }
 
     // Events that we want to update as soon as possible.  Return next time this should be called.
     private int SendFastRateEvents() {
-      // Handle startup and shutdown. And do not fire any events until the page has loaded and had a chance to
-      // register its event handlers.
-      if (Overlay == null || Overlay.Renderer == null || Overlay.Renderer.Browser == null || Overlay.Renderer.Browser.IsLoading) {
-        return kSlowTimerMilli;
-      }
-
       if (reset_notify_state_)
         notify_state_ = new NotifyState();
       reset_notify_state_ = false;
 
-      // Loading dance:
+      // Loading dance: (TODO: @ngld: this needs to be updated)
```

Comment:
Just for my own understanding, I would definitely appreciate this being updated.  :)

As far as I can tell, there's no more DOMContentReady, so it seems like C# just sends events as soon as it can, and whenever the JS side gets itself sorted, then it sends back a cactbotRequestState to re-send everything again?  That actually seems really reasonable and probably a bit more robust.

Reply:
Overlays only receive events once they've subscribed to them (`addOverlayHandler` handles that logic so you probably won't notice this part).

The current flow works like this:
* OverlayPlugin loads addons and initializes event sources.
* OverlayPlugin loads its configuration.
* Event sources are told to load their configuration and start (`LoadConfig` and `Start` are called).
* Overlays are initialised and the browser instances are started. At this points the overlays start loading.
* At some point the overlay's JavaScript is executed and `OverlayPluginApi` is injected. This order isn't deterministic and depends on what the ACT process is doing at that point in time. During startup the `OverlayPluginApi` is usually injected after the overlay is done loading while an overlay that's reloaded or loaded later on will see the `OverlayPluginApi` before the page has loaded.
* The overlay JavaScript sets up the initial event handlers and calls the `cactbotLoadUser` handler through `getUserConfigLocation`. These actions are queued by the JS implementation in `common.js` until `OverlayPluginApi` (or the WebSocket) is available. Once it is, the event subscriptions and handler calls are transmitted.
* OverlayPlugin stores the event subscriptions and executes the C# handlers which in this case means `FetchUserFiles` is called. That method loads the user files and returns them. The result is now transmitted back to the overlay that called the handler and the `Promise` in JS is resolved with the result.
* `getUserConfigLocation` processes the received information and calls the passed callback. This constructs the overlay specific objects and registers additional event handlers. Finally, the `cactbotRequestState` handler is called.
* OverlayPlugin processes the new event subscriptions and executes the `cactbotRequestState` handler.
* The next time `SendFastRateEvents` is called, it resets `notify_state_` (since the previous handler set `reset_notify_state_` to `true`) which causes it to dispatch all state events again. These events are now dispatched to all subscribed overlays. However, this means that overlays can receive state events multiple times during startup. If the user has three Cactbot overlays, all of them will call `cactbotRequestState` and thus cause this to happen one to three times depending on their timing. This shouldn't cause any issues but it's a waste of CPU cycles.
  Since this only happens during startup, it's probably not worth fixing though. Not sure.

Some overlays behave slightly different from the above explanation. Raidboss for example loads data files in addition to the listed steps. I think it's even loading them twice since `raidboss.js` loads the data files for `gTimelineController` and `popup-text.js` requests them again for its own purposes.

Generally, I've tried to follow two basic patterns with this overlay API:
* If the overlay needs information (and only once) before it can proceed (user files, data files, etc.), it calls a handler which returns exactly that information. With `Promises` this is easy enough to write (and would be even better if I could use `async` / `await` but I doubt that the old OverlayPlugin supports that) and it has the benefit that this information is only delivered to the overlay that requested the information.
* Event handlers are only attached once they're "ready". That means that all of the objects the event handler uses and all of the necessary data have been initialised and loaded.